fetcher/fetcher.go

```diff
@@ -0,0 +1,237 @@
+	f := &Fetcher{
+		im:          im,
+		cache:       make(map[string]*fetchData, numTxs),
+		keysToFetch: make(map[string][]ids.ID),
+		txnsToFetch: make(map[ids.ID]*sync.WaitGroup, numTxs),
+		fetchable:   make(chan *task),
+		stop:        make(chan struct{}),
+		numTxs:      numTxs,
+	}
+	for i := 0; i < concurrency; i++ {
+		f.wg.Add(1)
+		go f.runWorker()
+	}
+	return f
+}
+
+// Workers fetch individual keys
+func (f *Fetcher) runWorker() {
+	defer f.wg.Done()
+	for {
+		select {
+		case t, ok := <-f.fetchable:
+			if !ok {
+				return
+			}
+
+			// Allows concurrent reads to cache.
+			if exists := f.has(t.key); exists {
+				f.updateDependencies(t.key)
+				continue
+			}
+
+			// Doesn't exist in cache, check if we were the first to
+			// request the key.
+			f.keyLock.RLock()
+			firstTx := f.keysToFetch[t.key][0]
+			f.keyLock.RUnlock()
+			if firstTx != t.id {
+				t.wg.Wait()
+				continue
+			}
+
+			// Fetch from disk once if key isn't already in cache
+			v, err := f.im.GetValue(t.ctx, []byte(t.key))
+			if errors.Is(err, database.ErrNotFound) {
+				f.updateCache(t.key, nil, false, 0)
+				f.updateDependencies(t.key)
+				continue
+			} else if err != nil {
+				f.stopOnce.Do(func() {
+					f.err = err
+					close(f.stop)
+				})
+				return
+			}
+
+			// We verify that the [NumChunks] is already less than the number
+			// added on the write path, so we don't need to do so again here.
+			numChunks, ok := keys.NumChunks(v)
+			if !ok {
+				f.stopOnce.Do(func() {
+					f.err = ErrInvalidKeyValue
+					close(f.stop)
+				})
+				return
+			}
+
+			f.updateCache(t.key, v, true, numChunks)
+			f.updateDependencies(t.key)
+		case <-f.stop:
+			return
+		}
+	}
+}
+
+// Checks if a key is in the cache.
+func (f *Fetcher) has(k string) bool {
+	f.cacheLock.RLock()
+	defer f.cacheLock.RUnlock()
+	_, exists := f.cache[k]
+	return exists
+}
+
+// Puts a key that was fetched from disk into cache
+func (f *Fetcher) updateCache(k string, v []byte, exists bool, chunks uint16) {
+	f.cacheLock.Lock()
+	defer f.cacheLock.Unlock()
+	f.cache[k] = &fetchData{v, exists, chunks}
+}
+
+// After fetching a key, decrement the tx count that was waiting on the key.
+// This function is only called by the worker.
+func (f *Fetcher) updateDependencies(k string) {
+	f.keyLock.Lock()
+	defer f.keyLock.Unlock()
+
+	txIDs := f.keysToFetch[k]
+	f.txnLock.Lock()
+	for _, id := range txIDs {
+		f.txnsToFetch[id].Done() // Notify all other txs
+	}
+	f.txnLock.Unlock()
+	f.keysToFetch[k] = nil
+}
+
+// Lookup enqueues keys for the workers to fetch
+func (f *Fetcher) Lookup(ctx context.Context, txID ids.ID, stateKeys state.Keys) *sync.WaitGroup {
+	f.txnLock.Lock()
+	wg := &sync.WaitGroup{}
+	f.txnsToFetch[txID] = wg
+	f.txnsToFetch[txID].Add(len(stateKeys))
+	f.txnLock.Unlock()
+
+	f.keyLock.Lock()
+	tasks := make([]*task, 0, len(stateKeys))
+	for k := range stateKeys {
+		if _, ok := f.keysToFetch[k]; !ok {
+			f.keysToFetch[k] = make([]ids.ID, 0)
```

Comment:
While this approach is correct, I think it would be more efficient to avoid requeueing dependencies if the key has already been fetched. Seems odd to track them in that case.

Reply:
I think this would involve decrementing the `wg` count here or waiting to create the `wg` var until after this loop so we know what we are waiting on.

Reply:
You could handle this by using a `struct` as the value of `keysToFetch` that can communicate that the key was fetched to avoid more locks.